### DISCUSSION TITLE
Create 3 connection problem with PC via ethernet cable.

### DISCUSSION THREAD

**Discussion by omerdurmus61** [iRobotEducation/create3_docs#551] [Questions]
### How are you connecting to your Create 3?

Wi-Fi and USB Ethernet (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Intel Nuce Ubuntu 22.04

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.4

### Which RMW is your robot running?

FastRTPS

### Does your robot have an assigned namespace? If so, please share.

No

### Is the robot connected to a network? If so, what is the network type?

Office

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

No

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

Hello everyone, I have an Intel NUC mini PC on my Create 3 robot. There is an RPLIDAR A1 connected to the mini PC in my setup. I can access Create 3 topics from my laptop over WiFi. When I work with Nav2 and SLAM packages on my laptop (over WiFi), I encounter some errors due to delays. So, I decided to connect to the Create 3 via an Ethernet cable and run Nav2 and SLAM packages on the mini PC to avoid delays. On the mini PC, I navigated to the wired connection settings and changed the IPv4 method to manual, inputting an address of 192.168.186.3 and a netmask of 255.255.0.0. After completing that step, when I navigate to 192.168.186.2 in my browser, I can see the web server of the Create 3. However, when I run the 'ros2 topic list' command in the terminal, I couldn't see the Create 3 topics on my mini PC. Afterward, I used an XML file for FastRTPS RMW to disable multicasting. While this allowed me to see the Create 3 topics, I encountered issues such as being unable to run the launch file of my RPLIDAR and not being able to see the '/scan' topic on my mini PC. Additionally, I cannot see any of the topics I have published on the mini PC; I can only see the topics of the Create 3. Below is my XML file.

```
<?xml version="1.0" encoding="UTF-8" ?>
<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
   <participant profile_name="unicast_connection" is_default_profile="true">
       <rtps>
           <builtin>
               <metatrafficUnicastLocatorList>
                   <locator/>
               </metatrafficUnicastLocatorList>
               <initialPeersList>
                   <locator>
                       <udpv4>
                           <address>192.168.186.2</address>
                       </udpv4>
                   </locator>
               </initialPeersList>
           </builtin>
       </rtps>
   </participant>
</profiles>

```

**Comment by slowrunner** [comment]
@omerdurmus61 DISCLAIMER:  I am a Create3 user, not affiliated in any way with iRobot Education.  YMMV

**I SUSPECT** your configuration needs a FastDDS Discovery Server to tie all three parts together like this:

![ExamplePi4DiscoveryServer](https://github.com/iRobotEducation/create3_docs/assets/8950202/3a945015-f808-40f2-9226-5af42df881a1)

I could be totally wrong - I am quite new to this myself.


* ============= DISCOVERY SERVER SETUP =============

# Step by Step ( A GUESS - MY ENVIRONMENT IS DIFFERENT )

Starting point: Create3, ComputeBoard, and Desktop all running ROS 2 Humble FastDDS "Simple Discovery" - Full visibility on all three.  
(I do not know why with multicast on your devices could not discover each other)

6)  Setup Discovery Server to start on "Compute Board" (CB):
- CB: Add configs/super_client_configuration_file.xml with IP of 127.0.0.1:11811
  FILE REF: [super_client_configuration_file.xml]( https://fast-dds.docs.eprosima.com/en/latest/_downloads/9f9e92b14612364b742c8ecde24b3d24/super_client_configuration_file.xml )

```
<?xml version="1.0" encoding="UTF-8" ?>
<dds>
    <profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
        <participant profile_name="super_client_profile" is_default_profile="true">
            <rtps>
                <builtin>
                    <discovery_config>
                        <discoveryProtocol>SUPER_CLIENT</discoveryProtocol>
                        <discoveryServersList>
                            <RemoteServer prefix="44.53.00.5f.45.50.52.4f.53.49.4d.41">
                                <metatrafficUnicastLocatorList>
                                    <locator>
                                        <udpv4>
                                            <address>127.0.0.1</address>
                                            <port>11811</port>
                                        </udpv4>
                                    </locator>
                                </metatrafficUnicastLocatorList>
                            </RemoteServer>
                        </discoveryServersList>
                    </discovery_config>
                </builtin>
            </rtps>
        </participant>
    </profiles>
</dds>

```
- CB: Add to end of ~/.bashrc:

```
export ROS_DISCOVERY_SERVER="127.0.0.1:11811"
export FASTRTPS_DEFAULT_PROFILES_FILE=/home/ubuntu/configs/super_client_configuration_file.xml
```

- CB: Configure Discovery Server to run in background - (replace CB_wifi_IP with actual IPv4 x.x.x.x)

```
# echo -e "\n*** Start FastDDS Discovery Server"
echo '*** fastdds discovery -i 0 -l 192.168.186.3 -p 11811 -l CB_wifi_IP -p 11811 -l 127.0.0.1 -p 11811 &'
fasts discovery -i 0 -l 192.168.186.3 -p 11811 -l CB_wifi_IP -p 11811 -l 127.0.0.1 -p 11811 &
export ROS_DISCOVERY_SERVER="127.0.0.1:11811"
export FASTRTPS_DEFAULT_PROFILES_FILE=/home/ubuntu/configs/super_client_configuration_file.xml
```

7)  Create3: Enable Fast DDS discovery server in Create3 Application Configuration
    - check Enable Fast DDS discovery server? box
    - Enter 192.168.186.3:11811  (looks like it is there but need to enter it)
    - Click Save
    - Click Restart Application

8)  Configure desktop ROS CLI and any nodes to be "SuperClients" of discovery server:

- Desktop: Create file /home/ubuntu/configs/super_client_configuration_file.xml: (replace CB_WIFI_IP with actual IPv4 x.x.x.x)

```
<?xml version="1.0" encoding="UTF-8" ?>
<dds>
    <profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
        <participant profile_name="super_client_profile" is_default_profile="true">
            <rtps>
                <builtin>
                    <discovery_config>
                        <discoveryProtocol>SUPER_CLIENT</discoveryProtocol>
                        <discoveryServersList>
                            <RemoteServer prefix="44.53.00.5f.45.50.52.4f.53.49.4d.41">
                                <metatrafficUnicastLocatorList>
                                    <locator>
                                        <udpv4>
                                            <address>CB_WIFI_IP</address>
                                            <port>11811</port>
                                        </udpv4>
                                    </locator>
                                </metatrafficUnicastLocatorList>
                            </RemoteServer>
                        </discoveryServersList>
                    </discovery_config>
                </builtin>
            </rtps>
        </participant>
    </profiles>
</dds>


```
- Desktop: Add to ~/.bashrc:

```
export FASTRTPS_DEFAULT_PROFILES_FILE=/home/ubuntu/configs/super_client_configuration_file.xml
export ROS_DISCOVERY_SERVER=10.0.0.219:11811

```

8)  CB: Confirm Create3 topics visible on Compute Board (/dock_status, /odom, /ir_intensity)
    - Confirm immediately
    - Confirm after doing a warm boot of Pi5
    - Confirm after doing a storage/ship shutdown, boot of Create3 and ComputeBoard

9)  ComputeBoard: Confirm ComputeBoard topics seen by Create3 (publisth /cmd_vel, publish /dock)

10) Desktop: Confirm Create3 topics reach desktop (/battery_state, /dock_status) and rviz (/odom)

11) Desktop: Confirm ComputeBoard topics reach rviz (/mapUpdate, /oak/color/image_raw,  /scan)

12) Desktop: Confirm Desktop topics reach Create3 (pub /dock)
